### Importing Dependencies

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

import warnings
warnings.filterwarnings('ignore')

### Importing Data

In [3]:
train_df = pd.read_csv('train.csv')
test_df = pd.read_csv('test.csv')

In [4]:
store_df = pd.read_csv('store.csv')

In [5]:
train_df.head()

Store  DayOfWeek        Date  Sales  Customers  Open  Promo StateHoliday  \
0      1          5  2015-07-31   5263        555     1      1            0   
1      2          5  2015-07-31   6064        625     1      1            0   
2      3          5  2015-07-31   8314        821     1      1            0   
3      4          5  2015-07-31  13995       1498     1      1            0   
4      5          5  2015-07-31   4822        559     1      1            0   

   SchoolHoliday  
0              1  
1              1  
2              1  
3              1  
4              1

In [6]:
test_df.head()

Id  Store  DayOfWeek        Date  Open  Promo StateHoliday  SchoolHoliday
0   1      1          4  2015-09-17   1.0      1            0              0
1   2      3          4  2015-09-17   1.0      1            0              0
2   3      7          4  2015-09-17   1.0      1            0              0
3   4      8          4  2015-09-17   1.0      1            0              0
4   5      9          4  2015-09-17   1.0      1            0              0

In [7]:
store_df.head()

Store StoreType Assortment  CompetitionDistance  CompetitionOpenSinceMonth  \
0      1         c          a               1270.0                        9.0   
1      2         a          a                570.0                       11.0   
2      3         a          a              14130.0                       12.0   
3      4         c          c                620.0                        9.0   
4      5         a          a              29910.0                        4.0   

   CompetitionOpenSinceYear  Promo2  Promo2SinceWeek  Promo2SinceYear  \
0                    2008.0       0              NaN              NaN   
1                    2007.0       1             13.0           2010.0   
2                    2006.0       1             14.0           2011.0   
3                    2009.0       0              NaN              NaN   
4                    2015.0       0              NaN              NaN   

     PromoInterval  
0              NaN  
1  Jan,Apr,Jul,Oct  
2  Jan,Apr,Jul,Oct  
3              NaN  
4              NaN

In [8]:
print(f"Shape of train data: {train_df.shape}")
print(f"Shape of test  data: {test_df.shape}")
print(f"Shape of store data: {store_df.shape}")

Shape of train data: (1017209, 9)
Shape of test  data: (41088, 8)
Shape of store data: (1115, 10)


### Joining

In [9]:
# check if the stores in the train and store dataset align
train_df.Store.nunique() == store_df.Store.nunique()

True

In [11]:
df = train_df.merge(store_df, how='left', left_on=train_df.Store, right_on=store_df.Store)
df.head()

key_0  Store_x  DayOfWeek        Date  Sales  Customers  Open  Promo  \
0      1        1          5  2015-07-31   5263        555     1      1   
1      2        2          5  2015-07-31   6064        625     1      1   
2      3        3          5  2015-07-31   8314        821     1      1   
3      4        4          5  2015-07-31  13995       1498     1      1   
4      5        5          5  2015-07-31   4822        559     1      1   

  StateHoliday  SchoolHoliday  Store_y StoreType Assortment  \
0            0              1        1         c          a   
1            0              1        2         a          a   
2            0              1        3         a          a   
3            0              1        4         c          c   
4            0              1        5         a          a   

   CompetitionDistance  CompetitionOpenSinceMonth  CompetitionOpenSinceYear  \
0               1270.0                        9.0                    2008.0   
1                570.0                       11.0                    2007.0   
2              14130.0                       12.0                    2006.0   
3                620.0                        9.0                    2009.0   
4              29910.0                        4.0                    2015.0   

   Promo2  Promo2SinceWeek  Promo2SinceYear    PromoInterval  
0       0              NaN              NaN              NaN  
1       1             13.0           2010.0  Jan,Apr,Jul,Oct  
2       1             14.0           2011.0  Jan,Apr,Jul,Oct  
3       0              NaN              NaN              NaN  
4       0              NaN              NaN              NaN

In [12]:
df.drop(['key_0', 'Store_y'], axis=1, inplace=True)
df = df.rename(columns={'Store_x':'Store'})
df.head()

Store  DayOfWeek        Date  Sales  Customers  Open  Promo StateHoliday  \
0      1          5  2015-07-31   5263        555     1      1            0   
1      2          5  2015-07-31   6064        625     1      1            0   
2      3          5  2015-07-31   8314        821     1      1            0   
3      4          5  2015-07-31  13995       1498     1      1            0   
4      5          5  2015-07-31   4822        559     1      1            0   

   SchoolHoliday StoreType Assortment  CompetitionDistance  \
0              1         c          a               1270.0   
1              1         a          a                570.0   
2              1         a          a              14130.0   
3              1         c          c                620.0   
4              1         a          a              29910.0   

   CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
0                        9.0                    2008.0       0   
1                       11.0                    2007.0       1   
2                       12.0                    2006.0       1   
3                        9.0                    2009.0       0   
4                        4.0                    2015.0       0   

   Promo2SinceWeek  Promo2SinceYear    PromoInterval  
0              NaN              NaN              NaN  
1             13.0           2010.0  Jan,Apr,Jul,Oct  
2             14.0           2011.0  Jan,Apr,Jul,Oct  
3              NaN              NaN              NaN  
4              NaN              NaN              NaN

In [13]:
df.shape

(1017209, 18)

### Understanding the Data

In [14]:
round(df.describe().T, 2)

count     mean      std     min     25%  \
Store                      1017209.0   558.43   321.91     1.0   280.0   
DayOfWeek                  1017209.0     4.00     2.00     1.0     2.0   
Sales                      1017209.0  5773.82  3849.93     0.0  3727.0   
Customers                  1017209.0   633.15   464.41     0.0   405.0   
Open                       1017209.0     0.83     0.38     0.0     1.0   
Promo                      1017209.0     0.38     0.49     0.0     0.0   
SchoolHoliday              1017209.0     0.18     0.38     0.0     0.0   
CompetitionDistance        1014567.0  5430.09  7715.32    20.0   710.0   
CompetitionOpenSinceMonth   693861.0     7.22     3.21     1.0     4.0   
CompetitionOpenSinceYear    693861.0  2008.69     5.99  1900.0  2006.0   
Promo2                     1017209.0     0.50     0.50     0.0     0.0   
Promo2SinceWeek             509178.0    23.27    14.10     1.0    13.0   
Promo2SinceYear             509178.0  2011.75     1.66  2009.0  2011.0   

                              50%     75%      max  
Store                       558.0   838.0   1115.0  
DayOfWeek                     4.0     6.0      7.0  
Sales                      5744.0  7856.0  41551.0  
Customers                   609.0   837.0   7388.0  
Open                          1.0     1.0      1.0  
Promo                         0.0     1.0      1.0  
SchoolHoliday                 0.0     0.0      1.0  
CompetitionDistance        2330.0  6890.0  75860.0  
CompetitionOpenSinceMonth     8.0    10.0     12.0  
CompetitionOpenSinceYear   2010.0  2013.0   2015.0  
Promo2                        1.0     1.0      1.0  
Promo2SinceWeek              22.0    37.0     50.0  
Promo2SinceYear            2012.0  2013.0   2015.0

There are in total 1115 stores with sales feature having 3849.93 volatility and feature customers having 464.41 volatility with a mean of 5773.82 and 633.15 respectively.

In [15]:
train_df.duplicated().sum(), test_df.duplicated().sum()

(0, 0)

In [16]:
train_df.isnull().sum().sum(), test_df.isnull().sum().sum()

(0, 11)

There are no duplicates and Testing dataset has 11 null values.

In [18]:
print("Training data starts from: {}".format(train_df.Date.min()))
print("Training data end on: {}".format(train_df.Date.max()))
print()
print("Testing data starts from: {}".format(test_df.Date.min()))
print("Testing data end on: {}".format(test_df.Date.max()))

Training data starts from: 2013-01-01
Training data end on: 2015-07-31

Testing data starts from: 2015-08-01
Testing data end on: 2015-09-17
